sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -3371,14 +3371,505 @@ block_load(AccessorT lacc, simd_mask<1> pred, simd<T, N> pass_thru,
+/// Otherwise, the block stored operation is a NO-OP.
+///
+/// The parameter 'props' specifies the optional compile-time properties
+/// list. Only esimd::alignment property is used. Other properties are ignored.
+///
+/// Alignment: If \p props does not specify the 'alignment' property, then
+/// the default expected alignment is the minimally required (see (R1) below).
+///
+/// Restrictions - predicate imposed - temporary:
+/// R1: The \p byte_offset must be at least 4-byte aligned for 4-byte or smaller
+///     elements and 8-byte aligned for 8-byte elements.
+/// R2: The number of elements must be:
+///     for 8-byte data: 1, 2, 3, 4, 8, 16, 32(max for DG2), 64;
+///     for 4-byte data: 1, 2, 3, 4, 8, 16, 32, 64(max for DG2),
+///                      or 128(only if alignment is 8-bytes or more);
+///     for 2-byte data: 2, 4, 6, 8, 16, 32, 64, 128(max for DG2),
+///                      or 256(only if alignment is 8-bytes or more);
+///     for 1-byte data: 4, 8, 12, 16, 32, 64, 128, 256(max for DG2),
+///                      or 512(only if alignment is 8-bytes or more).
+/// R3: The target device must be DG2, PVC or newer GPU.
+template <typename T, int N, typename RegionTy = region1d_t<T, N, 1>,
+          typename PropertyListT =
+              ext::oneapi::experimental::detail::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT>>
+slm_block_store(uint32_t byte_offset, simd_view<T, RegionTy> vals,
+                simd_mask<1> pred, PropertyListT props = {}) {
+  slm_block_store<T, N>(byte_offset, vals.read(), pred, props);
+}
+
+/// void slm_block_store(uint32_t byte_offset,                   // (slm-bs-4)
+///                      simd_view<T, RegionTy> vals, props = {});
+/// Stores the vector \p vals to a contiguous memory block in SLM
+/// (Shared Local Memory) at the given \p byte_offset. The parameter 'props'
+/// specifies the optional compile-time properties list. Only esimd::alignment
+/// property is used. Other properties are ignored.
+///
+/// Alignment: If \p props does not specify the 'alignment' property, then
+/// the default expected alignment is 16-bytes to generate block_store
+/// instruction on all known target devices (Gen12, DG2, PVC, etc).
+/// On Gen12 (opposing to DG2 and PVC) the alignment smaller than 8-bytes
+/// is valid, but requires JIT compiler generating a slower SCATTER instead
+/// of faster BLOCK_STORE.
+/// !!! Passing \p byte_offset not aligned by 16-bytes and not specifying
+/// the actual alignment in \p props produces incorrect store results on Gen12.
+template <typename T, int N, typename RegionTy = region1d_t<T, N, 1>,
+          typename PropertyListT =
+              ext::oneapi::experimental::detail::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT>>
+slm_block_store(uint32_t byte_offset, simd_view<T, RegionTy> vals,
```

Comment:
I looked at this code one more time and probably found some problem here.
Can you tell why this code compiles and runs properly?

simd_view is the class with 2 template parameters `<BaseTy, RegionTy>`, where `BaseTy` is the base type for viewed objects, e.g. `simd_obj_impl`, not just `T`.

Reply:
This code https://github.com/intel/llvm/blob/a40009d2815fb61d54e3296d02a48acfee6d8a19/sycl/include/sycl/ext/intel/esimd/memory.hpp#L4341 is a bit similar, but it looks safer. There `Toffset' is not related to `T`. Reader may think that `Toffset` is either `uint32_t` or `uint64`, but in fact it is `simd_obj_impl`, right? If that is correct, then the comments to that function may be not correct (or misleading), but it should work correctly.

The function here uses simd_view<T, RegionTy>` and that connects/fixes `T` to scalar type - element of the returned value. How can `T` be deduced taking into account that it must a scalar/vectoriable type and `simd_obj_impl`?

Reply:
Sorry, it took me a bit to figure this out, I ended up having to dump the AST.
You are right, using `T` is wrong and that overload is never used. However, it works today because of the implicit conversion operator for simd_view to simd defined [here](https://github.com/intel/llvm/blob/462d6d4fe794f30c4ffe815f04cee7f72dd09064/sycl/include/sycl/ext/intel/esimd/detail/simd_view_impl.hpp#L118) and it ends up calling the normal simd one.

I think we have this problem for many of the new unified APIs and definitely need some follow up work to fix those.

Would you prefer we merge this PR as-is because the functions all work (but maybe for the wrong reason) and fix everything in a follow up PR, or would you prefer I fix the new simd_view overloads added in this PR now and deal with the other APIs in a follow up?

Reply:
Just had a discussion with @turinevgeny we were wondering if we should even have explicit simd_view overloads if we can rely on the conversion operator, maybe we can just delete all of those overloads?

Reply:
First I verified the all overloads of the current simd_view functions are called in the tests and that there is some runtime call that acutally hits those calls

Second I removed all of the simd_view overloads

Third I ran all of these tests on PVC and everything compiled successfully and passed.

Just pushed a commit removing the simd_view overloads.

Reply:
The fist of PRs adding simd_view overloads was here:
headers: https://github.com/intel/llvm/pull/7411
tests: https://github.com/intel/llvm-test-suite/pull/1385
There is no any tests there verifying passing offsets as `simd_view`. There was some reason for adding them - compilation errors, if I remember it correctly.
Need to understand the reasoning for adding simd_view overloads there and understand if it similar cases or different.

Reply:
Ok, taking into account that you added tests for simd_view args too.
It seems simd_view overload is needed in more special situations, not like those in block_store().